### ISSUE TITLE
test.py CUDA Assertion "index out of bounds"

### ISSUE THREAD

**Issue by khelkun** [MoyGcc/vid2avatar#18]
I had the `RuntimeError: element 0 of tensors does not require grad and does not have a grad_fn` while running `test.py` as reported in my #16 and @MoyGcc already answered in #5.

However downgrade to `pytorch-lightning==1.5.7` as suggested will prevent me to use `train.py` because I had to upgrade to `pytorch-lightning==1.9.5` to fix an error occuring during training epoch 20, please see [this post](https://github.com/MoyGcc/vid2avatar/issues/13#issuecomment-1551696277).

Whatever I'm more interested into running `test.py` and ultimately on my own  video. So I downgraded to `pytorch-lightning==1.5.7` and use the @MoyGcc  pre-trained checkpoint to run `test.py`. I got the memory error of #5, so I changed the [test.py num_sample](https://github.com/MoyGcc/vid2avatar/blob/main/code/confs/dataset/video.yaml#L36) from `-1` to `256`. This gave me the following error:
```
KeyError: Caught KeyError in DataLoader worker process 0.
Original Traceback (most recent call last):
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/torch/utils/data/_utils/worker.py", line 287, in _worker_loop
    data = fetcher.fetch(index)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 49, in fetch
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/torch/utils/data/_utils/fetch.py", line 49, in <listcomp>
    data = [self.dataset[idx] for idx in possibly_batched_index]
  File "/home/vizua/vid2avatar/code/lib/datasets/dataset.py", line 186, in __getitem__
    "img_size": images["img_size"]
KeyError: 'img_size'
```
So I fixed [this line of code in dataset.py](https://github.com/MoyGcc/vid2avatar/blob/main/code/lib/datasets/dataset.py#L101) like this:
```
#            images = {"rgb": samples["rgb"].astype(np.float32)}
            images = {
                "rgb": samples["rgb"].astype(np.float32),
                "img_size": img_size
            }
```
Now `test.py` runs but it fails on a CUDA assertion that I'm unable to fix:
```
$ python test.py
test.py:9: UserWarning:
The version_base parameter is not specified.
Please specify a compatability version level, or None.
Will assume defaults for version 1.1
  @hydra.main(config_path="confs", config_name="base")
/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/_internal/hydra.py:127: UserWarning: Future Hydra versions will no longer change working directory at job runtime by default.
See https://hydra.cc/docs/1.2/upgrades/1.1_to_1.2/changes_to_job_working_dir/ for more information.
  configure_logging=with_log_configuration,
Global seed set to 42
Working dir: /home/vizua/vid2avatar/outputs/Video/parkinglot
GPU available: True, used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
Restoring states from the checkpoint path at checkpoints/epoch=6299-loss=0.01887552998960018.ckpt
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Loaded model weights from checkpoint at checkpoints/epoch=6299-loss=0.01887552998960018.ckpt
Testing: 0it [00:00, ?it/s]/home/vizua/vid2avatar/code/v2a_model.py:219: UserWarning: __floordiv__ is deprecated, and its behavior will change in a future version of pytorch. It currently rounds toward 0 (like the 'trunc' function NOT 'floor'). This results in incorrect rounding for negative values. To keep the current behavior, use torch.div(a, b, rounding_mode='trunc'), or for actual floor division, use torch.div(a, b, rounding_mode='floor').
  1) // pixel_per_batch
mise.MISE(32, 4, 0):
points shape: 35937
points: []
/home/vizua/vid2avatar/code/lib/utils/meshing.py:49: FutureWarning: marching_cubes_lewiner is deprecated in favor of marching_cubes. marching_cubes_lewiner will be removed in version 0.19
  level=level_set)
/pytorch/aten/src/ATen/native/cuda/IndexKernel.cu:93: operator(): block: [6,0,0], thread: [96,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/IndexKernel.cu:93: operator(): block: [6,0,0], thread: [97,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.

.................................................................................

/pytorch/aten/src/ATen/native/cuda/IndexKernel.cu:93: operator(): block: [3,0,0], thread: [93,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/IndexKernel.cu:93: operator(): block: [3,0,0], thread: [94,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
/pytorch/aten/src/ATen/native/cuda/IndexKernel.cu:93: operator(): block: [3,0,0], thread: [95,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
Error executing job with overrides: []
Traceback (most recent call last):
  File "test.py", line 39, in <module>
    main()
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/main.py", line 99, in decorated_main
    config_name=config_name,
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/_internal/utils.py", line 401, in _run_hydra
    overrides=overrides,
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/_internal/utils.py", line 458, in _run_app
    lambda: hydra.run(
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/_internal/utils.py", line 223, in run_and_report
    raise ex
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/_internal/utils.py", line 220, in run_and_report
    return func()
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/_internal/utils.py", line 461, in <lambda>
    overrides=overrides,
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/_internal/hydra.py", line 132, in run
    _ = ret.return_value
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/core/utils.py", line 260, in return_value
    raise self._return_value
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/hydra/core/utils.py", line 186, in run_job
    ret.return_value = task_function(task_cfg)
  File "test.py", line 36, in main
    trainer.test(model, testset, ckpt_path=checkpoint)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 907, in test
    return self._call_and_handle_interrupt(self._test_impl, model, dataloaders, ckpt_path, verbose, datamodule)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 683, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 950, in _test_impl
    results = self._run(model, ckpt_path=self.tested_ckpt_path)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1195, in _run
    self._dispatch()
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1271, in _dispatch
    self.training_type_plugin.start_evaluating(self)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 206, in start_evaluating
    self._results = trainer.run_stage()
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1282, in run_stage
    return self._run_evaluate()
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py", line 1330, in _run_evaluate
    eval_loop_results = self._evaluation_loop.run()
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/loops/dataloader/evaluation_loop.py", line 110, in advance
    dl_outputs = self.epoch_loop.run(dataloader, dataloader_idx, dl_max_batches, self.num_dataloaders)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/loops/base.py", line 145, in run
    self.advance(*args, **kwargs)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py", line 122, in advance
    output = self._evaluation_step(batch, batch_idx, dataloader_idx)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py", line 213, in _evaluation_step
    output = self.trainer.accelerator.test_step(step_kwargs)
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/accelerators/accelerator.py", line 244, in test_step
    return self.training_type_plugin.test_step(*step_kwargs.values())
  File "/home/vizua/anaconda3/envs/v2a/lib/python3.7/site-packages/pytorch_lightning/plugins/training_type/training_type_plugin.py", line 222, in test_step
    return self.model.test_step(*args, **kwargs)
  File "/home/vizua/vid2avatar/code/v2a_model.py", line 250, in test_step
    batch_inputs = {"uv": inputs["uv"][:, indices],
RuntimeError: CUDA error: device-side assert triggered
Testing:   0%|          | 0/42 [00:06<?, ?it/s]
```

N.B: I have the exact same issue with owned trained model at epoch 800 with `pytorch-lightning==1.9.5`

**Comment by MoyGcc** [comment]
Hi, thank you for updating such a detailed bug report. Indeed, I think you should be able to run this repo by strictly following the dependency requirement as listed [here](https://github.com/MoyGcc/vid2avatar/blob/main/requirement.txt). But never mind, as long as you get it run, I think it should be fine. 

For your current issue, I would suggest changing the [pixel_per_batch](https://github.com/MoyGcc/vid2avatar/blob/main/code/confs/dataset/video.yaml#LL37C5-L37C20) attribute instead of [num_sample](https://github.com/MoyGcc/vid2avatar/blob/main/code/confs/dataset/video.yaml#L36) in order to get the test run on GPUs with less VRAM. I actually need to remove these unnecessary parameters in the config file (which were used during development). Sorry about the confusion.

**Comment by khelkun** [comment]
As you suggested I put back num_sample to -1 and decreased pisel_per_batch:
```
    num_sample : -1
    pixel_per_batch: 512
```
N.B: please also note that I decreased the `worker` from `8` to `4` because some warning suggested this to me when I first tried to train.  
Now `test.py` is running so thank you!!! It will probably take some long time to complete but I'll let you know if everything goes fine and if I'm able to run on my custom video.


> I think you should be able to run this repo by strictly following the dependency requirement as listed [here](https://github.com/MoyGcc/vid2avatar/blob/main/requirement.txt)

Sorry, I actually messed up a bit in the repository issues those last days. In #16 (that I closed myself) I mentioned #13 where I explained why I had to upgrade my dependencies [in this post](https://github.com/MoyGcc/vid2avatar/issues/13#issuecomment-1551696277) (and the 2 previous ones).

**Comment by khelkun** [comment]
`train.py` ran sucessfuly so I got the outputs. I'll close the current issue but have more questions on #19 if you have time @MoyGcc
